rasa/core/featurizers/tracker_featurizers.py

```diff
@@ -1044,13 +1040,13 @@ def training_states_labels_and_entities(
             ):
 
                 if self.remove_duplicates:
-                    hashed = self._hash_example(tracker, states, label)
+                    hashed = self._hash_example(states, label)
                     if hashed in hashed_examples:
                         continue
                     hashed_examples.add(hashed)
 
                 # Store all positive labels associated with a training state.
-                state_hash = self._hash_example(tracker, states)
+                state_hash = self._hash_example(states)
                 state_hash_to_label_list_instances[state_hash].append(label)
 
                 example_states.append(states)
```

Comment:
```suggestion

                if self.remove_duplicates:
                    hashed = self._hash_example(states, label)
                    if hashed in hashed_examples:
                        continue
                    hashed_examples.add(hashed)

                # Store all positive labels associated with a training state.
                state_hash = self._hash_example(states)
                state_hash_to_label_list_instances[state_hash].add(label[0])

                # Only add unique example states unless `remove_duplicates` is `False`.
                if not self.remove_duplicates or state_hash not in seen_states:
                    example_states.append(states)
                    example_entities.append(entities)
                    seen_states.add(state_hash)

                pbar.set_postfix({f"# {self.LABEL_NAME}": f"{len(example_states):d}"})

        # Collect positive labels for each state example.
        example_labels = [
            list(state_hash_to_label_list_instances[self._hash_example(state)])
            for state in example_states
        ]

        self._remove_user_text_if_intent(example_states)

        logger.debug(
            f"Created {len(example_states)} {self.LABEL_NAME} examples."
        )

        return example_states, example_labels, example_entities
```

(I can't extend the suggestion after line 1056, but this would replace everything from 1041 on in the function.)

What you have from lines 1058 on works, but IMO it seems wasteful of resources to (a) create both a large list of possibly redundant data and then a pruned list when we could just create the pruned list and (b) create `example_labels` when it isn't being used for anything.

The above suggestion would create pruned example states in one pass and simplify the creation of the pruned example labels later. Not necessary to do but I thought if we have a little more time since we have a release blocking bug we could simplify this code a bit.

This would also mean moving the creation of `seen_states` up to line 1022 and converting state_hash_to_label_list_instances to hold a set instead of a list of length 1 lists (and a name change).

If you don't do this suggestion, I still think you should remove `example_labels` since its not doing anything.